src/exec/use_pty/pipe.rs

```diff
@@ -222,15 +360,70 @@ impl<R: Read, W: Write> Buffer<R, W> {
 
     /// Flush this buffer, ensuring that all the contents of its internal buffer are written.
     fn flush(&mut self, write: &mut W) -> io::Result<()> {
-        // This is the busy section of the buffer.
-        let buffer = &self.buffer[self.start..self.end];
-
-        // Write the complete busy section to `write`.
-        write.write_all(buffer)?;
+        // Don't remove anything if the buffer is empty.
+        if self.is_empty() {
+            return write.flush();
+        }
 
-        // If we were able to write all the busy section, we can mark the whole buffer as free.
-        self.start = 0;
-        self.end = 0;
+        // We need to find all the sub-buffers with data.
+        if self.head < self.tail {
```

Comment:
having the helpers methods I mentioned above would simplify this:
``` rust
let mut wrote_something = false;
for _ in 0..2 {
    let chunk = self.ring_buffer.next_readable_chunk();
    if chunk.is_empty() { break }
    write.write_all(chunk)?;
    wrote_something = true;
    self.ring_buffer.advance_head(chunk.len());
}
if wrote_something { write.flush()?; }
```

Reply:
having helper methods and a separate `RingBuffer` struct would also make this easier to test. for example:
``` rust
let mut rb = RingBuffer::new();
rb.head = 0;
rb.tail = self.buffer.len() - 1;
assert!(rb.next_writable_chunk().is_empty());
```
that is, you could turn the scenarios in your diagrams into tests.